File: plant.cpp
#include "plant.h"

// constructor & destructor
Plant::Plant(int ip, int op, int dl)
    :inputPin{ip}, outputPin{op}, drynessLimit{dl} {
        moistureLevel = 0;
    }
Plant::Plant(const Plant &source)
    :inputPin{source.inputPin}, outputPin{source.outputPin}, moistureLevel{source.moistureLevel}, drynessLimit{source.drynessLimit} {}
Plant::~Plant() {}

// getters & setters
int Plant::getInput() {
    return inputPin;
}
int Plant::getOutput() {
    return outputPin;
}
int Plant::getMoistureLevel() {
    return moistureLevel;
}
int Plant::getDrynessLimit() {
    return drynessLimit;
}

void Plant::setInput(int in) {
    pinMode(in, INPUT);
}
void Plant::setOutput(int in) {
    pinMode(in, OUTPUT);
}
void Plant::setMoistureLevel() {
    moistureLevel = analogRead(this->inputPin);
}
void Plant::setDrynessLimit(int in) {
    drynessLimit = in;
}

//methods
void Plant::waterIfNeeded() {
    if (this->getMoistureLevel() > this->getDrynessLimit()) { // if moisture levels are below desired level
        Serial.print("Starting watering\n");
        Serial.print(this->getOutput());
        digitalWrite(this->getOutput(), LOW); // start pump
        delay(5000); // 5 seconds
        digitalWrite(this->getOutput(), HIGH); // stop pump
        Serial.print("Stopped watering\n");
    }
}

File: plant.py
import time


class Plant:
    def __init__(self, name: str, dl: int) -> None:
        self.name = name
        self.drynessLimit = dl
        self.moistureLevel: int = 0
    
    # getters/setters
    def get_name(self) -> str:
        return self.name

    def get_moisture_level(self) -> int:
        return self.moistureLevel

    def get_dryness_limit(self) -> int:
        return self.drynessLimit
    
    def set_moisture_level(self, ml) -> None:
        self.moistureLevel = ml

    def set_dryness_limit(self, dl) -> None:
        self.drynessLimit = dl

    # methods
    def water_if_needed(self) -> None:
        if self.moistureLevel > self.drynessLimit:
            print("Starting watering")
            # send start command to arduino
            time.sleep(1)
            # send stop command to arduino
            print("Stopped watering")
